File: LeetCode/Medium/0875_Koko_Eating_Bananas/Solution.py
from math import *
from typing import *

## 2022-08-29 (binary search)
# 
# Given that we need to find a value within a defined range (koko can eat at a speed of
# 1 to X bananas/hr), we can utilize binary search to cut down the number of values we
# need to search through.
# 
class Solution:
    def timeToEatAll(self, piles, speed) -> int:
        return sum([ceil(pile/speed) for pile in piles])
        
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        lo, hi = 1, max(piles)
        cur_min = hi
        while lo <= hi:
            mid = lo + (hi-lo)//2
            duration = self.timeToEatAll(piles, mid)
            
            if duration <= h:
                cur_min = mid
                hi = mid-1
            else:
                lo = mid+1
        
        return cur_min

File: LeetCode/Medium/0875_Koko_Eating_Bananas/Solution.cpp
// 2022-12-26 (binary search)
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        /*
            input: n piles of bananas, and h hours
            output: minimum int K that we can finish all bananas within h hours

            notes:
            - we kinda need to simulate it, but we need to find the best val within a range
            - most liekly binary search since finding within a range

            approach:
            - find min and max values within piles
            - set up binary search where the middle value would be how many we eat
              per hour
            - 'simulate' eating by running through the pile and seeing if our hours are sufficient
            - lowest value we finish eating thru will be our return
        */
        int smallest = 1;
        int largest = *max_element(piles.begin(), piles.end());
        while (smallest < largest) {
            int eat = smallest + (largest-smallest)/2;
            int hoursLeft = h;
            for (int pile : piles) {
                hoursLeft -= ceil((1.0 * pile)/eat);
            }
            if (hoursLeft >= 0) {
                largest = eat;
            } else {
                smallest = eat+1;
            }
        }
        return smallest;
    }
};